### PR TITLE
Minor improvement for snappy decompression in parquet

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description
Minor improvement for snappy decompression in parquet

## Additional context and related issues
These changes are based on existing code from ORC reader

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java [hunk 1]
```diff
@@ -72,9 +72,14 @@ public static Slice decompress(CompressionCodec codec, Slice input, int uncompre
 
     private static Slice decompressSnappy(Slice input, int uncompressedSize)
     {
-        byte[] buffer = new byte[uncompressedSize];
-        decompress(new SnappyDecompressor(), input, 0, input.length(), buffer, 0);
-        return wrappedBuffer(buffer);
+        // Snappy decompressor is more efficient if there's at least a long's worth of extra space
+        // in the output buffer
+        byte[] buffer = new byte[uncompressedSize + SIZE_OF_LONG];
+        int decompressedBytes = decompress(new SnappyDecompressor(), input, 0, input.length(), buffer, 0);
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:83]: nit: `decompressedBytes` -> `actualUncompressedSize`?

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java [hunk 2]
```diff
@@ -72,9 +72,14 @@ public static Slice decompress(CompressionCodec codec, Slice input, int uncompre
 
     private static Slice decompressSnappy(Slice input, int uncompressedSize)
     {
-        byte[] buffer = new byte[uncompressedSize];
-        decompress(new SnappyDecompressor(), input, 0, input.length(), buffer, 0);
-        return wrappedBuffer(buffer);
+        // Snappy decompressor is more efficient if there's at least a long's worth of extra space
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:80]: why?

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java [hunk 3]
```diff
@@ -70,34 +70,42 @@ public static Slice decompress(CompressionCodec codec, Slice input, int uncompre
         throw new ParquetCorruptionException("Codec not supported in Parquet: " + codec);
     }
 
-    private static Slice decompressSnappy(Slice input, int uncompressedSize)
+    public interface OutputBuffer
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:73]: nit: `OutputBufferProvider`?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java [hunk 4]
```diff
@@ -37,23 +40,27 @@
...
 {
     private final CompressionCodec codec;
     private final boolean hasOnlyDictionaryEncodedPages;
     private final boolean hasNoNulls;
     private final PeekingIterator<Page> compressedPages;
+    private final LocalMemoryContext decompressionBufferMemoryUsage;
 
     private boolean dictionaryAlreadyRead;
     private int dataPageReadCount;
+    private byte[] decompressorOutputBuffer;
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:55]: Can reusing of buffer interfere with block laziness (e.g. produced `Blocks` become stalled)?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java [hunk 5]
```diff
@@ -103,8 +112,11 @@ public DataPage readPage()
         dataPageReadCount++;
         try {
             if (compressedPage instanceof DataPageV1 dataPageV1) {
+                if (codec == UNCOMPRESSED) {
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:115]: separate commit?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java [hunk 6]
```diff
@@ -175,13 +187,19 @@ public void skipNextPage()
...
+        checkArgument(dictionaryAlreadyRead, "Dictionary has to be read first");
     }
 
-    private void verifyDictionaryPageRead()
+    private OutputBuffer createOutputBuffer()
     {
-        checkArgument(dictionaryAlreadyRead, "Dictionary has to be read first");
+        return size -> {
+            if (decompressorOutputBuffer == null || size > decompressorOutputBuffer.length) {
+                decompressorOutputBuffer = new byte[size];
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:199]: Can there be a rogue `size` that would dramatically increase reader memory usage? Should buffer be shrink after a while?

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java [hunk 7]
```diff
@@ -70,34 +70,42 @@ public static Slice decompress(CompressionCodec codec, Slice input, int uncompre
...
+    {
+        byte[] initialize(int size);
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:75]: Shouldn't it be more like `aquire/release` semantics? When buffer is released? Add a comment please

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java (hunk 6 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:199]: This wouldn't have been problem as page sizes are stable within a column chunk (every writer is aiming for some size and values count threshold).

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java (hunk 7 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:75]: Buffer was tied to PageReader which is tied to per-row group column reader which gets cleaned up whenever we reach of end of a row group

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:75]: Sematics here is similar to the OutputBuffer in orc reader

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java (hunk 2 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/ParquetCompressionUtils.java:75]: See https://github.com/airlift/aircompressor/blob/master/src/main/java/io/airlift/compress/snappy/SnappyRawDecompressor.java#L78
Code change is copied from OrcSnappyDecompressor

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java [hunk 8]
```diff
@@ -103,6 +103,9 @@ public DataPage readPage()
         dataPageReadCount++;
         try {
             if (compressedPage instanceof DataPageV1 dataPageV1) {
+                if (!arePagesCompressed()) {
+                    return dataPageV1;
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:107]: is `dataPageV1` reused in any way?

**Review by sopel39 — APPROVED** [review]

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java (hunk 8 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/PageReader.java:107]: Existing code is also logically the same, `decompress` will just give back the input slice without modifications for uncompressed case.

**Pull request merged**